drivers/xen/gnttab.c

```diff
@@ -0,0 +1,333 @@
+{
+	int ret;
+	void *page_addr;
+	struct xen_remove_from_physmap rfpm;
+
+	page_addr = k_aligned_alloc(XEN_PAGE_SIZE, XEN_PAGE_SIZE);
+	if (!page_addr) {
+		LOG_WRN("Failed to allocate memory for gnttab page!\n");
+		return NULL;
+	}
+
+	rfpm.domid = DOMID_SELF;
+	rfpm.gpfn = xen_virt_to_gfn(page_addr);
+
+	/*
+	 * GNTTABOP_map_grant_ref will simply replace the entry in the P2M
+	 * and not release any RAM that may have been associated with
+	 * page_addr, so we release this memory before mapping.
+	 */
+	ret = HYPERVISOR_memory_op(XENMEM_remove_from_physmap, &rfpm);
+	if (ret) {
+		LOG_WRN("Failed to remove gnttab page from physmap, ret = %d\n", ret);
+		return NULL;
+	}
+
+	return page_addr;
+}
+
+void gnttab_put_page(void *page_addr)
+{
+	int ret, nr_extents = 1;
+	struct xen_memory_reservation reservation;
+	xen_pfn_t page = xen_virt_to_gfn(page_addr);
+
+	/*
+	 * After unmapping there will be a 4Kb holes in address space
+	 * at 'page_addr' positions. To keep it contiguous and be able
+	 * to return such addresses to memory allocator we need to
+	 * populate memory on unmapped positions here.
+	 */
+	memset(&reservation, 0, sizeof(reservation));
+	reservation.domid = DOMID_SELF;
+	reservation.extent_order = 0;
+	reservation.nr_extents = nr_extents;
+	set_xen_guest_handle(reservation.extent_start, &page);
+
+	ret = HYPERVISOR_memory_op(XENMEM_populate_physmap, &reservation);
+	if (ret != nr_extents) {
+		LOG_WRN("failed to populate physmap on gfn = 0x%llx, ret = %d\n",
+			page, ret);
+		return;
+	}
+
+	k_free(page_addr);
+}
+
+int gnttab_map_refs(struct gnttab_map_grant_ref *map_ops, unsigned int count)
+{
+	int i, ret;
+
+	ret = HYPERVISOR_grant_table_op(GNTTABOP_map_grant_ref, map_ops, count);
+	if (ret) {
+		return ret;
+	}
+
+	for (i = 0; i < count; i++) {
+		switch (map_ops[i].status) {
+		case GNTST_no_device_space:
+			LOG_WRN("map_grant_ref failed, no device space for page #%d\n", i);
+			break;
+
+		case GNTST_eagain:
+			/* Operation not done; need to try again */
+			gop_eagain_retry(GNTTABOP_map_grant_ref, &map_ops[i]);
+			/* Need to re-check status for current page */
+			i--;
+
+			break;
+
+		default:
+			break;
+		}
+	}
+
+	return 0;
+}
+
+int gnttab_unmap_refs(struct gnttab_map_grant_ref *unmap_ops, unsigned int count)
+{
+	return HYPERVISOR_grant_table_op(GNTTABOP_unmap_grant_ref, unmap_ops, count);
+}
+
+
+static const char * const gnttab_error_msgs[] = GNTTABOP_error_msgs;
+
+const char *gnttabop_error(int16_t status)
+{
+	status = -status;
+	if (status < 0 || (uint16_t) status >= ARRAY_SIZE(gnttab_error_msgs)) {
+		return "bad status";
+	} else {
+		return gnttab_error_msgs[status];
+	}
+}
+
+static int gnttab_init(const struct device *d)
+{
+	grant_ref_t gref;
+	struct xen_add_to_physmap xatp;
+	struct gnttab_setup_table setup;
+	xen_pfn_t frames[NR_GRANT_FRAMES];
+	int rc = 0, i;
+
+	/* Will be taken/given during gnt_refs allocation/release */
+	k_sem_init(&gnttab.sem, 0, NR_GRANT_ENTRIES - 1);
```

Comment:
@luca-fancellu, sorry for late reply. Initial value should still be 0, init loop will set it to right value after initializing all grefs. But limit should be `NR_GRANT_ENTRIES - GNTTAB_NR_RESERVED_ENTRIES` instead of `NR_GRANT_ENTRIES - 1` (1 was reserved for gref[0], where first free gref is holded). Thank you for advice!

Reply:
@firscity ok great, once this is fixed, I think this is ready to go. I will try to ping reviewers for that so it can be merged soon.

Reply:
I think that to make the CI happy, you need to put at least one line of content in the commit message of this one: xenvm: doc: update documentation for Xen grant tables

Reply:
Oh, thanks, I will try this now.